packages/react-instantsearch-core/src/core/indexUtils.js

```diff
@@ -161,44 +167,99 @@ function getNamespaceAndAttributeName(id) {
   return { namespace, attributeName };
 }
 
-// eslint-disable-next-line max-params
+function hasRefinements({
+  context,
+  indexId,
+  namespace,
+  attributeName,
+  id,
+  searchState,
+}) {
+  const multiIndex = hasMultipleIndices(context);
+
+  if (multiIndex && namespace) {
+    return (
+      searchState &&
+      searchState.indices &&
+      searchState.indices[indexId] &&
+      searchState.indices[indexId][namespace] &&
+      searchState.indices[indexId][namespace].hasOwnProperty(attributeName)
+    );
+  }
+
+  if (multiIndex) {
+    return (
+      searchState &&
+      searchState.indices &&
+      searchState.indices[indexId] &&
+      searchState.indices[indexId].hasOwnProperty(id)
+    );
+  }
+
+  if (namespace) {
+    return (
+      searchState &&
+      searchState[namespace] &&
+      searchState[namespace].hasOwnProperty(attributeName)
+    );
+  }
+
+  return searchState && searchState.hasOwnProperty(id);
+}
+
+function getRefinements({
+  context,
+  namespace,
+  searchState,
+  indexId,
+  attributeName,
+  id,
+}) {
+  const multiIndex = hasMultipleIndices(context);
+
+  if (multiIndex && namespace) {
+    return (
+      searchState.indices[indexId][namespace] &&
+      searchState.indices[indexId][namespace][attributeName]
+    );
+  }
+  if (multiIndex) {
+    return searchState.indices[indexId] && searchState.indices[indexId][id];
+  }
+  if (namespace) {
+    return searchState[namespace] && searchState[namespace][attributeName];
+  }
+
+  return searchState && searchState[id];
+}
+
 export function getCurrentRefinementValue(
   props,
   searchState,
   context,
   id,
-  defaultValue,
-  refinementsCallback = x => x
+  defaultValue
 ) {
   const indexId = getIndexId(context);
   const { namespace, attributeName } = getNamespaceAndAttributeName(id);
-  const refinements =
-    (hasMultipleIndices(context) &&
-      searchState.indices &&
-      namespace &&
-      searchState.indices[`${indexId}`] &&
-      has(searchState.indices[`${indexId}`][namespace], `${attributeName}`)) ||
-    (hasMultipleIndices(context) &&
-      searchState.indices &&
-      has(searchState, `indices.${indexId}.${id}`)) ||
-    (!hasMultipleIndices(context) &&
-      namespace &&
-      has(searchState[namespace], attributeName)) ||
-    (!hasMultipleIndices(context) && has(searchState, id));
-  if (refinements) {
-    let currentRefinement;
-
-    if (hasMultipleIndices(context)) {
-      currentRefinement = namespace
-        ? get(searchState.indices[`${indexId}`][namespace], attributeName)
-        : get(searchState.indices[indexId], id);
-    } else {
-      currentRefinement = namespace
-        ? get(searchState[namespace], attributeName)
-        : get(searchState, id);
-    }
+  const hasRefinementsValue = hasRefinements({
+    context,
```

Comment:
Inside both functions, the `context` is only used to detect whether or not we are in a multi-index context. We can hoist the call at this level and provide the value directly to the functions rather than delegate the computation.